File: src/main/java/com/zippeykeys/praisebe/util/EnumUtil.java
package com.zippeykeys.praisebe.util;

import java.util.Arrays;

import lombok.experimental.UtilityClass;
import lombok.var;


@UtilityClass
public class EnumUtil{
    public static <T extends Enum<T>> Enum<T> byIndex(T enumType, int index){
        return getValues(enumType)[index];
    }

    public static <T extends Enum<T>> int getSize(T enumVar){
        return getValues(enumVar).length;
    }

    @SuppressWarnings("unchecked")
    public static <T extends Enum<T>> Enum<T>[] getValues(T enumVar){
        return getValues((Class<T>) enumVar.getClass());
    }

    public static <T extends Enum<T>> Enum<T>[] getValues(Class<? extends T> enumVar){
        return enumVar.getEnumConstants();
    }

    @SafeVarargs
    public static <T extends Enum<T>> int getUniqueIndex(T... enums){
        int result = enums[0].ordinal();
        T[] x = Arrays.copyOfRange(enums, 1, enums.length);
        for(var enumVar : x){
            result = enumVar.ordinal() + getSize(enumVar) * result;
        }
        return result;
    }
}

File: src/main/kotlin/com/zippeykeys/praisebe/util/EnumUtil.kt
package com.zippeykeys.praisebe.util

import java.util.EnumMap


inline operator fun <reified T : Enum<T>> T.get(index: Int): T {
    return enumValues<T>()[index]
}

fun enumSize(clazz:Class<*>): Int {
    return clazz.enumConstants.size
}

fun enumIndices(clazz:Class<*>): IntRange {
    return clazz.enumConstants.indices
}

inline fun <reified K : Enum<K>, V> List<V>.toEnumMap(): EnumMap<K, V> {
    return enumValues<K>().zip(this)
            .toMap(EnumMap(K::class.java))
}